### ISSUE TITLE
feat: add orbax checkpointing

### ISSUE THREAD

**Issue by fbanelli** [antonioterpin/synthpix#208]
**Is your feature request related to a problem? Please describe.**
Currently, training resumes are not fully deterministic and checkpointing the data pipeline is complex. The root cause is a **split state** between CPU and GPU:
1.  **CPU (Grain)**: Manages file reading and shuffling logic.
2.  **GPU (Sampler)**: Manages its own JAX `PRNGKey` and internal iteration counters (e.g., `batches_per_flow`).

If we only checkpoint Grain, the Sampler's RNG resets on resume, causing it to generate different images for the same flow field. Synchronizing these two independent states via manual serialization is fragile and error-prone.

**Describe the solution you'd like**
I propose a unified strategy that refines our randomness handling to enable simple, robust checkpointing via Orbax.

### 1. Refined Randomness: Grain as the Source of Truth
To make the Sampler restorable, we must make it **stateless** regarding the "source of entropy". 
- **Grain Path**: We add a `RandomMapTransform` to the Grain pipeline. For every record, Grain provides a deterministic **integer seed** (`jax_seed`).
- **Sampler Path**: The Sampler converts the integer into a JAX key (`jax.random.PRNGKey(jax_seed)`) and then uses `jax.random.fold_in(key, rep_idx)` to derive unique noise for each variation. 
- **Legacy Path**: To maintain backward compatibility, the Sampler falls back to an internal `_rng` if no seed is provided by the scheduler.

### 2. Modern Orbax Checkpointing
By moving the RNG "anchor" to Grain, we can use the **Modern Orbax Args-based API** to save the entire pipeline state in one atomic transaction.

#### Saving
```python
mngr.save(
    step=global_step,
    args=ocp.args.Composite(
        # Handles binary Grain iterator state (records, shuffle position)
        data=grain.PyGrainCheckpointSave(grain_iterator),
        # Handles Sampler metadata (step counter, legacy _rng) as a PyTree
        sampler=ocp.args.StandardSave({"step": sampler._step, "_rng": sampler._rng}),
    )
)
```

#### Restoring
```python
restored = mngr.restore(
    step=ckpt_step,
    args=ocp.args.Composite(
        data=grain.PyGrainCheckpointRestore(grain_iterator),
        sampler=ocp.args.StandardRestore({"step": 0, "_rng": initial_rng}),
    )
)
# The sampler is restored perfectly because both the record and the counter 
# (which hashes the noise) are synchronized.
sampler._step = restored.sampler["step"]
sampler._rng = restored.sampler["_rng"]
```

**Why this is necessary**
This change turns the randomness problem from a "state synchronization" problem into a "derivation" problem. By hashing the Grain-provided seed with the Sampler-provided step, we guarantee that **variation #3 of flow #401** is identical every time, regardless of whether it's a first run or a resume from checkpoint.

**Additional context**
- This follows the official [Grain-Orbax integration guide](https://google-grain.readthedocs.io/en/latest/tutorials/data_loader_tutorial.html#integration-with-orbax).
